### PR TITLE
OSF-3425 Prevent line breaks in file extension

### PR DESCRIPTION
**Pull request by ag4ml**
# Purpose

Currently, if the window is resized on a file page, there is a possibility of an "awkward" line break with the file name. This means that part of the file extension could be split and a .docx could be interpreted as a .doc.

b92f05f0-1e76-11e5-8495-087cf0d98fb0

# Changes

* Add a new CSS rule to the file-view-page.css file using the "file-ext" id
* Modify the view_file.mako file so that it splits the file name into two parts: with and without extension
* Apply the CSS id to the extension part

### CODE REVIEW

**Comment by samanehsan** [comment]
Can you resolve the merge conflicts? Pulling from the develop branch should also fix these failing tests (they don't seem to be related to the change you made).

**Comment by samanehsan** [comment]
Pass complete @ag4ml! :whale:

File: website/templates/project/view_file.mako [hunk 1]
```diff
@@ -6,12 +6,22 @@
...
-    <h2 class="break-word">
-      ${file_name | h}
-      % if file_revision:
-        <small>&nbsp;${file_revision | h}</small>
-      % endif
-    </h2>
+     <h2 class="break-word">
+       ## Split file name into two parts: with and without extension
+       <%
+        extStart = file_name.rfind(".")
```
Comment (samanehsan) [inline website/templates/project/view_file.mako]: If the file does not have an extension, but it does have a "." in the title, whatever comes after the "." is treated like an extension and the line breaks:  
screen shot 2015-10-26 at 4 24 31 pm  

screen shot 2015-10-26 at 4 24 41 pm


 Is there a more reliable way to get the file extension?

**Comment by ag4ml** [comment]
Hmmm...using the Python OS module seems to be more reliable, but I'm a little lost on how to figure out if a filename contains an extension in the first place. Maybe there is a solution out there with regular expressions. I have a feeling it can get complicated.

**Comment by ag4ml** [comment]
Oh wait, certain characters (and spaces) aren't allowed in file extensions so I can just check if the string that was extracted contains any of these illegal characters.

**Comment by ag4ml** [comment]
Now, it checks if the file name is POSIX portable. Because there's so many file systems with their own file naming conventions, it's debatable on what the standard actually should be.

**Comment by ag4ml** [comment]
https://en.wikipedia.org/wiki/Filename#Comparison_of_filename_limitations

**Comment by samanehsan** [comment]
Thanks for making those changes @ag4ml. Done with code review 2 :octopus:

File: website/templates/project/view_file.mako [hunk 2]
```diff
@@ -8,7 +8,20 @@
 <div class="row">
   <div class="col-sm-5">
     <h2 class="break-word">
-      ${file_name | h}
+      ## Split file name into two parts: with and without extension
+      <%
+       import os
+       part1, part2 = os.path.splitext(file_name)
+       ## POSIX standard for chars in a fully portable filename
+       validFileExtension = all(i.isalnum() or i in ['.', '_', '-'] for i in part2)
```
Comment (sloria) [inline website/templates/project/view_file.mako]: As per COSDev, use snake_case for python variable names.

File: website/templates/project/view_file.mako [hunk 3]
```diff
@@ -8,7 +8,20 @@
...
+       if validFileExtension:
```
Comment (sloria) [inline website/templates/project/view_file.mako]: @ag4ml Can you explain why this is necessary?

File: website/templates/project/view_file.mako [hunk 4]
```diff
@@ -8,7 +8,20 @@
...
+        file_name_title, file_name_ext = part1, part2
```
Comment (sloria) [inline website/templates/project/view_file.mako]: The `part1` and `part2` variables feel unnecessary.

```python
base_name, file_ext = os.path.splitext(file_name)
is_valid_ext =  ...
if not is_valid_ext:
    file_ext = ''
```

**Comment by sloria** [comment]
This is a lot of logic to go into the templates. I think it would be more appropriate to put this logic in the view. @ag4ml Could you refactor this?

**Comment by sloria** [comment]
Pass finished.

File: website/templates/project/view_file.mako (hunk 3 above)
Comment (ag4ml) [inline website/templates/project/view_file.mako]: When a file name doesn't have an extension, but the file name contains a ".", like for "Really long title 10.26.15 important stuff", the "15 important stuff" was treated as an actual extension when it wasn't supposed to be.
The code in line 16 checks if the result from os.path.splitext() is really supposed to be an extension.

File: website/templates/project/view_file.mako (hunk 4 above)
Comment (ag4ml) [inline website/templates/project/view_file.mako]: Yeah, I agree.

**Comment by ag4ml** [comment]
Could you expand more upon how I'd go about refactoring this (or point me somewhere that discusses this)? Where would the code go exactly?

File: website/templates/project/view_file.mako [hunk 5]
```diff
@@ -8,7 +8,8 @@
 <div class="row">
   <div class="col-sm-5">
     <h2 class="break-word">
-      ${file_name | h}
+      ## Split file name into two parts: with and without extension
+      ${file_name_title | h}<span id="file-ext">${file_name_ext | h}</span>
```
Comment (sloria) [inline website/templates/project/view_file.mako]: Much better :+1:

**Pull request merged**